### PR TITLE
api: add `rev` param to api calls

### PR DESCRIPTION
**Pull request by Suor**

### CODE REVIEW

File: dvc/repo/__init__.py [hunk 1]
```diff
@@ -455,9 +456,15 @@ def open(self, path, remote=None, mode="r", encoding=None):
...
 
+        cache_file = self.cache.local.checksum_to_path_info(out.checksum)
+        cache_file = fspath_py35(cache_file)
+
         with self.state:
             cache_info = out.get_used_cache(remote=remote)
             self.cloud.pull(cache_info, remote=remote)
+            if not os.path.exists(cache_file):
+                raise DvcException(
+                    "Can't find {path} neither locally nor on remote"
```
Comment (shcheklein) [inline dvc/repo/__init__.py]: are we testing this?

missing `.format(path` ...?

Comment (shcheklein) [reply dvc/repo/__init__.py]: what is the motivation to throw this here instead of relying on the `_open` itself? To unify the behavior across Pythons?

Comment (shcheklein) [reply dvc/repo/__init__.py]: also, does it make sense to make a specific exception `DvcFileNotFoundException`? And put more info into the exception message - like remote info we tried to pull it from.

File: dvc/api.py [hunk 2]
```diff
@@ -45,29 +46,31 @@ def __getattr__(self, name):
...
+    with _make_repo(repo, rev=rev) as _repo:
         abspath = os.path.join(_repo.root_dir, path)
         with _repo.open(
             abspath, remote=remote, mode=mode, encoding=encoding
         ) as fd:
             yield fd
 
 
-def read(path, repo=None, remote=None, mode="r", encoding=None):
+def read(path, repo=None, rev=None, remote=None, mode="r", encoding=None):
```
Comment (shcheklein) [inline dvc/api.py]: btw, what will happen if I call it for a directory?

File: dvc/repo/__init__.py (hunk 1 above)
Comment (Suor) [reply dvc/repo/__init__.py]: When built-in `open()` raises here the error message looks quite cryptic, it is some random cache path not existing, while the usual real reason is it absent on remote. `self.cloud.pull()` already shows a warning with what is not found, so no need to add something about it here. 

I will write a test.

File: dvc/api.py (hunk 2 above)
Comment (Suor) [reply dvc/api.py]: It should work as long as `Repo(some_dir, rev=rev)` works, which is problematic, but that is a separate issue - #2201.

Comment (Suor) [reply dvc/api.py]: An example where that won't work - remote configuration differs in a branch. Now a config file is always read from a actual fs, not git tree.

**Comment by Suor** [comment]
`rev` for local repos is disabled with assert until #2201 is resolved. The rest is addressed.

**Comment by efiop** [comment]
@Suor windows tests are failing

**Comment by efiop** [comment]
Might be a false-negative, restarted.

**Comment by Suor** [comment]
@efiop sometimes I randomly can't remove tmp repo dir, sometimes it's something unrelated.

**Pull request merged**